src/InfluxDB2/WriteRetry.php

```diff
@@ -0,0 +1,143 @@
+<?php
+
+namespace InfluxDB2;
+
+use GuzzleHttp\Exception\ConnectException;
+
+/**
+ * Exponential random write retry.
+ */
+class WriteRetry
+{
+    private $maxRetries;
+    private $retryInterval;
+    private $maxRetryDelay;
+    private $exponentialBase;
+    private $jitterInterval;
+    private $maxRetryTime;
+    private $retryTimout;
+    /**
+     * @var mixed|string
+     */
+    private $logFile;
+
+    /**
+     * WriteRetry constructor.
+     *
+     * @param int $maxRetries max number of retries when write fails
+     * @param int $retryInterval number of milliseconds to retry unsuccessful write,
+     *      The retry interval is used when the InfluxDB server does not specify "Retry-After" header.
+     * @param int $maxRetryDelay maximum delay when retrying write in milliseconds
+     * @param int $exponentialBase the base for the exponential retry delay, the next delay is computed using
+     *              random exponential backoff as a random value within the interval
+     *              ``retryInterval * exponentialBase^(attempts-1)`` and
+     *              ``retryInterval * exponentialBase^(attempts)``.
+     *              Example for ``retryInterval=5000, exponentialBase=2, maxRetryDelay=125000, total=5``
+     *              Retry delays are random distributed values within the ranges of
+     *              ``[5000-10000, 10000-20000, 20000-40000, 40000-80000, 80000-125000]``
+     *
+     * @param int $maxRetryTime maximum total time when retrying write in milliseconds
+     * @param int $jitterInterval the number of milliseconds before the data is written increased by a random amount
+     * @param string $logFile logfile
+     */
+    public function __construct(
+        int $maxRetries = 5,
+        int $retryInterval = 5000,
+        int $maxRetryDelay = 125000,
+        int $exponentialBase = 2,
+        int $maxRetryTime = 180000,
+        int $jitterInterval = 0,
+        string $logFile = "php://output"
+    ) {
+        $this->maxRetries = $maxRetries;
+        $this->retryInterval = $retryInterval;
+        $this->maxRetryDelay = $maxRetryDelay;
+        $this->maxRetryTime = $maxRetryTime;
+        $this->exponentialBase = $exponentialBase;
+        $this->jitterInterval = $jitterInterval;
+        $this->logFile = $logFile;
+
+        //retry timout
+        $this->retryTimout = microtime(true) * 1000 + $maxRetryTime;
+    }
+
+    /**
+     * @throws ApiException
+     */
+    public function retry($callable, $attempts = 0)
+    {
+        try {
+            return call_user_func($callable);
+        } catch (ApiException $e) {
+            $error = $e->getResponseBody() ?? $e->getMessage();
+
+            if (!$this->isRetryable($e)) {
+                throw $e;
+            }
+            $attempts++;
+            if ($attempts > $this->maxRetries || !$this->isRetryable($e)) {
```

Comment:
`!$this->isRetryable($e)` is useless because it is already checked at line `74`

```suggestion
            if ($attempts > $this->maxRetries) {
```